File: javascript/dnd-character/dnd-character.js
function d6() {
  const [diceSides, diceMinValue] = [6, 1];
  return diceMinValue + Math.floor(Math.random() * diceSides);
}

export const abilityModifier = (abilityScore) => {
  if (abilityScore < 3) {
    throw new RangeError("Ability scores must be at least 3");
  }
  if (abilityScore > 18) {
    throw new RangeError("Ability scores can be at most 18");
  }
  return Math.floor(abilityScore / 2) - 5;
};

export class Character {

  #strength;
  #dexterity;
  #constituition;
  #intelligence;
  #wisdom;
  #charisma;

  static rollAbility() {
    return [0, 1, 2, 3]
      .map(d6)
      .sort()
      .slice(1)
      .reduce((sum, v) => sum + v);
  }

  get strength() {
    this.#strength = this.#strength ?? Character.rollAbility();
    return this.#strength;
  }

  get dexterity() {
    this.#dexterity = this.#dexterity ?? Character.rollAbility();
    return this.#dexterity;
  }

  get constitution() {
    this.#constituition = this.#constituition ?? Character.rollAbility();
    return this.#constituition;
  }

  get intelligence() {
    this.#intelligence = this.#intelligence ?? Character.rollAbility();
    return this.#intelligence;
  }

  get wisdom() {
    this.#wisdom = this.#wisdom ?? Character.rollAbility();
    return this.#wisdom;
  }

  get charisma() {
    this.#charisma = this.#charisma ?? Character.rollAbility();
    return this.#charisma;
  }

  get hitpoints() {
    return 10 + abilityModifier(this.#constituition);
  }
}

File: java/dnd-character/src/main/java/DnDCharacter.java
import java.util.List;
import java.util.Random;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

class DnDCharacter {

    private int strength,
            constitution,
            dexterity,
            intelligence,
            wisdom,
            charisma;

    private final static Random random = new Random();

    public DnDCharacter() {
        this.strength = this.generateAbilityScore();
        this.dexterity = this.generateAbilityScore();
        this.constitution = this.generateAbilityScore();
        this.intelligence = this.generateAbilityScore();
        this.wisdom = this.generateAbilityScore();
        this.charisma = this.generateAbilityScore();
    }

    int ability(List<Integer> scores) {
        final int dropLowest = 1, keepHighest = 3;
        return scores
                .stream()
                .sorted()
                .skip(dropLowest)
                .limit(keepHighest)
                .reduce(0, Integer::sum);
    }

    List<Integer> rollDice() {
        final int dicesRolleds = 4, diceFaces = 6, minFace = 1;

        return IntStream
                .range(0, dicesRolleds)
                .map((_ignored) -> random.nextInt(minFace, diceFaces))
                .boxed()
                .collect(Collectors.toList());
    }

    int modifier(int input) {
        final int baseScore = 10;
        return Math.floorDiv(input - baseScore, 2);
    }

    int getStrength() {
        return this.strength;
    }

    int getDexterity() {
        return this.dexterity;
    }

    int getConstitution() {
        return this.constitution;
    }

    int getIntelligence() {
        return this.intelligence;
    }

    int getWisdom() {
        return this.wisdom;
    }

    int getCharisma() {
        return this.charisma;
    }

    int getHitpoints() {
        final int baseHp = 10;
        return baseHp + this.modifier(this.constitution);
    }

    private int generateAbilityScore() {
        return this.ability(this.rollDice());
    }

}
